Fix for Ranger 2.4 HMSA startup issue

xBis7 opened the pull request:
Ranger used to fail during startup due to an issue with the `jersey-json` dependency.

Ranger-2.4 had a dependency on Hadoop 3.3.0 which had a dependency on `com.sun.jersey:jersey-json`. That way ranger was getting `com.sun.jersey:jersey-json` from its hadoop dependency.

In Hadoop 3.3.6, the package was replaced for the `jersey-json` dep and instead of `com.sun.jersey:jersey-json`, the dependency was now accessed from `com.github.pjfanning:jersey-json`.

By upgrading Ranger's Hadoop dependency to 3.3.6, we replaced `com.sun.jersey:jersey-json` with `com.github.pjfanning:jersey-json` but Ranger was still having a dependency on `com.sun.jersey:jersey-json` on modules where there was no hadoop dep and it had to get `jersey-json` itself.

That was the cause of the startup issue. To fix it, we need to exclude `com.github.pjfanning:jersey-json` from all hadoop dependencies and manually add it wherever the module doesn't get it from another Ranger module.

I've compared the dependency trees for the current branch and branch `ranger-2.4` and I've made sure that all the packages that used to use `jersey-json`, still have it available only once.

xBis7 committed:
replace jersey-json dep package from hadoop with com.sun.jersey

xBis7 committed:
replace com.github.pjfanning with com.sun.jersey in the rest of the modules

xBis7 commented:
@mladjan-gadzic @GeorgeJahad can you take a look at this PR? Can you also verify that it fixes the startup issue locally for you?

mladjan-gadzic commented:
> @mladjan-gadzic @GeorgeJahad can you take a look at this PR? Can you also verify that it fixes the startup issue locally for you?

I don't have my Intel MB at hand right now and don't wanna risk local env. I will test it as soon as I can.

xBis7 commented:
> I don't have my Intel MB at hand right now and don't wanna risk local env. I will test it as soon as I can.

ok, don't worry. Whenever you get some time.

Here are some steps for testing
* get the branch
* run a clean install on Ranger so that it downloads all the new dependencies
  * `mvn clean compile package install --batch-mode -DskipTests -DskipShade`
* run this script from the helper scripts repo
  * `./setup/setup_ranger_with_rebuild.sh /home/projects/`
  * It deletes all tarballs and docker images, forces a rebuild and starts the docker env
* after the script finishes, check the UI

GeorgeJahad commented:
This builds and works fine for me but I don't quite understand how you decided when to add an exclusion, when to add an inclusion, and when to add both?

xBis7 commented:
I checked the dependency tree for branch `ranger-2.4` and then for `ranger-2.4-with-hmsa` and I compared the outputs.

I've put the outputs in gists

* `ranger-2.4`
https://gist.github.com/xBis7/44a52a57cde0ff2e5385e24cb8e37854

* `ranger-2.4-with-hmsa` + changes (intermediate state)
https://gist.github.com/xBis7/0bed5009f1a0780fa3a9170a5aee83e8

* final dependency tree of the PR branch
https://gist.github.com/xBis7/80513ac9da1b8b3dada332b12c6d219c

I checked the logs for `ranger-2.4`, I searched for `jersey-json` and I noted down all the packages that had this dependency. Almost every module used to get it from Hadoop 3.3.0 and we can see that in every module it exists only once.

Because in Hadoop 3.3.6, the `jersey-json` package has been replaced from `com.sun.jersey` to `com.github.pjfanning`, now all Hadoop dependencies give us `com.github.pjfanning:jersey-json`.

I excluded `com.github.pjfanning:jersey-json` from all Hadoop dependencies. Now, we needed to somehow get `com.sun.jersey:jersey-json` and we needed it to appear only once. 

I manually included it in common ranger modules that get used throughout the project (e.g. plugins-common) and then checked the dependencies one by one for the other modules. 

In Ranger modules where there was a dependency that included another Ranger module that was bringing in `com.sun.jersey:jersey-json` I just let it be and moved on. In modules where it was originally available but no other dependency was bringing it in, I added it manually.

xBis7 commented:
We can avoid all the changes in this PR by modifying only the parent pom.

New PR: https://github.com/xBis7/ranger/pull/3

Closing this one.

xBis7 commented:
We can avoid all the changes in this PR by modifying only the parent pom.

New PR: https://github.com/xBis7/ranger/pull/3

Closing this one.
